README.md
[Также доступно на русском языке](README_RU.md)

# Variscite Dart

[![Pub Package](https://img.shields.io/pub/v/variscite_dart.svg)](https://pub.dev/packages/variscite_dart)
[![GitHub License](https://img.shields.io/badge/License-MIT-blue.svg)](https://raw.githubusercontent.com/uslashvlad/variscite_dart/master/LICENSE)

**Part of "Variscite" project. Here is also [API](https://github.com/uSlashVlad/variscite_api) and [client app](https://github.com/uSlashVlad/variscite_mobile)**

Library for [Dart](https://dart.dev) programming language for work with [Variscite API](https://github.com/uSlashVlad/variscite_api). Written poorly with Dart and uses this libraries:

- **dio** - library for working with HTTP, creating and sending requests to API
- **geojson_vi** - library for working with GeoJSON objects, namely for constructing them from JSON
- **equatable** - auxiliary library for more convenient comparation of API objects

Library implements all API endpoints of version 0.3, you can find list of all methods in file _lib/src/api.dart_

# Examples

Group creation and login:

```dart
import 'package:variscite_dart/variscite_dart.dart';

void main() async {
    // At the start, it is necessary to initialize VarisciteApi
    // class. It performs all functions, related to API
    final api = VarisciteApi();

    // Creation of group with name "New group" and admin password
    // "qwerty"
    final group = await api.createNewGroup(
        GroupCreationBody(name: 'New group', passcode: 'qwerty'),
    );

    // Logging in group with name "Admin" and passcode "qwerty"
    // If passcode doesn't exists, user won't have admins rights
    final user = await api.loginUsingInviteCode(
        group.inviteCode,
        UserLoginBody(name: 'Admin', passcode: 'qwerty'),
    );

    // Setting up user token, token is needed for any of farther
    // actions, related to group
    api.setToken(user.token);
}
```

Retrieving group information and list of users:

```dart
import 'package:variscite_dart/variscite_dart.dart';

void main() async {
    final api = VarisciteApi(token: 'super.secret.token');

    // Retrieving group information: name, invite code, id
    final groupInfo = await api.getCurrentGroupInfo();

    // Retrieving list of users, each element contains name,
    // rights information (admin or not) and user id
    final users = await api.getUsersList();
}
```

Working with geographical structures:

```dart
import 'package:variscite_dart/variscite_dart.dart';

void main() async {
    final api = VarisciteApi(token: 'super.secret.token');

    // Retrieving list of all structures, each element contains
    // structure' id and id of user created it, additional fields
    // (they can be edited separately) and most important -
    // GeoJSON object (it has FeatureCollection type)
    final structures = await api.getAllStructures();

    // Creating Map object that contains GeoJSON object
    final feature = {
        'type': 'FeatureCollection',
        'features': [
            {
                'type': 'Feature',
                'properties': {},
                'geometry': {
                'type': 'Polygon',
                    'coordinates': [
                        [
                            [34.453125, 54.007768],
                            [40.78125, 54.007768],
                            [40.78125, 57.326521],
                            [34.453125, 57.326521],
                            [34.453125, 54.007768],
                        ],
                    ],
                },
            },
        ],
    };
    // Transforming Map<String, dynamic> into GeoJSONFeatureCollection
    final collectionGeoJSON = GeoJSONFeatureCollection.fromMap(feature);

    // Creating structure from GeoJSON object, storing it into
    // database
    final newStruct = await api.addNewStructure(collectionGeoJSON);
}
```

Working with geolocation:

```dart
import 'package:variscite_dart/variscite_dart.dart';

void main() async {
    final api = VarisciteApi(token: 'super.secret.token');

    // Retrieving geolocation of all users with open geolocation
    // in group
    final allLocations = await api.getAllGeolocation();
    // If you need to remove geolocation of current user
    // (user that sended request), you can achive it by passing
    // "excludeUser: true" argument
    final almostAllLocations = await api.getAllGeolocation(excludeUser: true);

    // Updating geolocation of current user
    await api.editCurrentUserGeolocation(GeolocationPosition(
        latitude: 37.478860,
        longitude: 55.670802,
    ));
}
```

Specifying address of your own API:

```dart
import 'package:variscite_dart/variscite_dart.dart';

void main() async {
    // In this case reqiests will be sended to https://example.com
    final api = VarisciteApi(serviceUrl: 'https://example.com');
}
```


CHANGELOG.md
## 0.1.4

* Geolocation fromMap() fix
* Explicit "exclude_user" parameter
* Added posibility to remove token from API handler

## 0.1.3

* Made it possible to get token of API handler
* Added option to exclude location of current user

## 0.1.2

* Added "hasToken" property to API handler

## 0.1.1

* Better example
* Made it possible to change url of Variscite API service

## 0.1.0

Initial version, all features of Variscite API v0.3.0 was implemented


lib/src/api.dart
import 'network.dart';
import 'models.dart';
import 'consts.dart';

import 'package:geojson_vi/geojson_vi.dart';

class VarisciteApi {
  late RequestsHandler _req;
  String? _token;

  String? get userToken => _token;
  bool get hasToken => _token != null;

  VarisciteApi({String? serviceUrl, String? token}) {
    if (token != null) {
      _token = token;
      _req = RequestsHandler(
        serviceUrl ?? defaultTestServerUrl,
        headers: _genTokenHeader(token),
      );
    } else {
      _req = RequestsHandler(
        serviceUrl ?? defaultTestServerUrl,
      );
    }
  }

  void setToken(String token) {
    _token = token;
    _req = RequestsHandler(
      defaultTestServerUrl,
      headers: _genTokenHeader(token),
    );
  }

  void removeToken() {
    _token = null;
    _req = RequestsHandler(defaultTestServerUrl);
  }

  Map<String, String> _genTokenHeader(String token) {
    return {'Authorization': 'Bearer $token'};
  }

  Future<GroupCreationOutput> createNewGroup(GroupCreationBody body) async {
    final res = await _req.postRequest('/groups', body.toMap());
    return GroupCreationOutput.fromMap(res.data);
  }

  Future<UserLoginOutput> loginUsingInviteCode(
    String inviteCode,
    UserLoginBody body,
  ) async {
    final res = await _req.postRequest('/groups/$inviteCode', body.toMap());
    return UserLoginOutput.fromMap(res.data);
  }

  Future<GroupInfo> getCurrentGroupInfo() async {
    final res = await _req.getRequest('/groups/my');
    return GroupInfo.fromMap(res.data);
  }

  Future<void> deleteCurrentGroup() async {
    await _req.deleteRequest('/groups/my');
  }

  Future<List<UserInfo>> getUsersList() async {
    final res = await _req.getRequest('/groups/my/users');
    final userMaps = res.data as List;
    return userMaps.map((map) => UserInfo.fromMap(map)).toList();
  }

  Future<UserInfo> getUser(String userId) async {
    final res = await _req.getRequest('/groups/my/users/$userId');
    return UserInfo.fromMap(res.data);
  }

  Future<void> deleteUser(String userId) async {
    await _req.deleteRequest('/groups/my/users/$userId');
  }

  Future<UserInfo> getCurrentUserInfo() async {
    final res = await _req.getRequest('/groups/my/users/me');
    return UserInfo.fromMap(res.data);
  }

  Future<void> deleteCurrentUser() async {
    await _req.deleteRequest('/groups/my/users/me');
  }

  Future<List<GeoStruct>> getAllStructures() async {
    final res = await _req.getRequest('/structures');
    final structMaps = res.data as List;
    return structMaps.map((map) => GeoStruct.fromMap(map)).toList();
  }

  Future<GeoStruct> addNewStructure(GeoJSONFeatureCollection geojson) async {
    final res = await _req.postRequest('/structures', geojson.toMap());
    return GeoStruct.fromMap(res.data);
  }

  Future<GeoStruct> getStructure(String structId) async {
    final res = await _req.getRequest('/structures/$structId');
    return GeoStruct.fromMap(res.data);
  }

  Future<GeoStruct> editStructure(
    String structId,
    GeoJSONFeatureCollection geojson,
  ) async {
    final res = await _req.putRequest('/structures/$structId', geojson.toMap());
    return GeoStruct.fromMap(res.data);
  }

  Future<void> deleteStructure(String structId) async {
    await _req.deleteRequest('/structures/$structId');
  }

  Future<Map<String, dynamic>> getStructureFields(String structId) async {
    final res = await _req.getRequest('/structures/$structId/fields');
    return Map.castFrom(res.data);
  }

  Future<Map<String, dynamic>> postStructureFields(
    String structId,
    Map<String, dynamic> fields,
  ) async {
    final res = await _req.postRequest('/structures/$structId/fields', fields);
    return Map.castFrom(res.data);
  }

  Future<Map<String, dynamic>> deleteStructureFields(
    String structId,
    List<String> fields,
  ) async {
    final fieldsStr = fields.join(',');
    final res = await _req.deleteRequest('/structures/$structId/fields'
        '?fields=$fieldsStr');
    return Map.castFrom(res.data);
  }

  Future<List<UserGeolocationOutput>> getAllGeolocation(
      {bool excludeUser = false}) async {
    final res =
        await _req.getRequest('/location/all?exclude_user=$excludeUser');
    final locationMaps = res.data as List;
    return locationMaps
        .map((map) => UserGeolocationOutput.fromMap(map))
        .toList();
  }

  Future<GeolocationPosition> getUserGeolocation(String userId) async {
    final res = await _req.getRequest('/location/$userId');
    return GeolocationPosition.fromMap(res.data);
  }

  Future<GeolocationPosition> getCurrentUserGeolocation() async {
    final res = await _req.getRequest('/location/my');
    return GeolocationPosition.fromMap(res.data);
  }

  Future<GeolocationPosition> editCurrentUserGeolocation(
      GeolocationPosition position) async {
    final res = await _req.putRequest('/location/my', position.toMap());
    return GeolocationPosition.fromMap(res.data);
  }

  Future<void> deleteCurrentGeolocation() async {
    await _req.deleteRequest('/location/my');
  }
}


lib/src/models.dart
import 'package:equatable/equatable.dart';
import 'package:geojson_vi/geojson_vi.dart';

abstract class ApiModel extends Equatable {
  const ApiModel();
}

abstract class RequestBodyInterface {
  Map<String, dynamic> toMap();
}

// I can't do the same interface for responce body and fromMap() method
// since here is no way in dart to create static inheretible methods
// So you really should be careful with this method and with such classes

class GroupCreationBody extends ApiModel implements RequestBodyInterface {
  final String name;
  final String passcode;

  const GroupCreationBody({
    required this.name,
    required this.passcode,
  });

  @override
  List<Object?> get props => [name, passcode];

  @override
  Map<String, dynamic> toMap() => {
        'name': name,
        'passcode': passcode,
      };
}

class GroupCreationOutput extends ApiModel {
  final String id;
  final String inviteCode;

  const GroupCreationOutput({
    required this.id,
    required this.inviteCode,
  });

  @override
  List<Object?> get props => [id, inviteCode];

  static GroupCreationOutput fromMap(map) => GroupCreationOutput(
        id: map['id'],
        inviteCode: map['inviteCode'],
      );
}

class UserLoginBody extends ApiModel implements RequestBodyInterface {
  final String name;
  final String? passcode;

  const UserLoginBody({
    required this.name,
    this.passcode,
  });

  @override
  List<Object?> get props => [name, passcode];

  @override
  Map<String, dynamic> toMap() => {
        'name': name,
        'passcode': passcode,
      };
}

class UserLoginOutput extends ApiModel {
  final String id;
  final String token;

  const UserLoginOutput({
    required this.id,
    required this.token,
  });

  @override
  List<Object?> get props => [id, token];

  static UserLoginOutput fromMap(map) => UserLoginOutput(
        id: map['id'],
        token: map['token'],
      );
}

class GroupInfo extends ApiModel {
  final String id;
  final String name;
  final String inviteCode;

  const GroupInfo({
    required this.id,
    required this.name,
    required this.inviteCode,
  });

  @override
  List<Object?> get props => [id, name, inviteCode];

  static GroupInfo fromMap(map) => GroupInfo(
        id: map['id'],
        name: map['name'],
        inviteCode: map['inviteCode'],
      );
}

class UserInfo extends ApiModel {
  final String id;
  final String name;
  final bool isAdmin;

  const UserInfo({
    required this.id,
    required this.name,
    required this.isAdmin,
  });

  @override
  List<Object?> get props => [id, name, isAdmin];

  static UserInfo fromMap(map) => UserInfo(
        id: map['id'],
        name: map['name'],
        isAdmin: map['isAdmin'],
      );
}

class GeoStruct extends ApiModel {
  final String id;
  final String user;
  final Map<String, dynamic> fields;
  final GeoJSONFeatureCollection struct;

  const GeoStruct({
    required this.id,
    required this.user,
    required this.fields,
    required this.struct,
  });

  @override
  List<Object?> get props => [id, user, fields, struct];

  static GeoStruct fromMap(map) => GeoStruct(
        id: map['id'],
        user: map['user'],
        fields: Map.castFrom(map['fields']),
        struct: GeoJSONFeatureCollection.fromMap(Map.castFrom(map['struct'])),
      );
}

class UserGeolocationOutput extends ApiModel {
  final String user;
  final GeolocationPosition position;

  const UserGeolocationOutput({
    required this.user,
    required this.position,
  });

  @override
  List<Object?> get props => [user, position];

  static UserGeolocationOutput fromMap(map) => UserGeolocationOutput(
        user: map['user'],
        position: GeolocationPosition.fromMap(map['position']),
      );
}

class GeolocationPosition extends ApiModel implements RequestBodyInterface {
  final double latitude;
  final double longitude;

  const GeolocationPosition({
    required this.latitude,
    required this.longitude,
  });

  @override
  List<Object?> get props => [latitude, longitude];

  @override
  Map<String, dynamic> toMap() => {
        'latitude': latitude,
        'longitude': longitude,
      };

  static GeolocationPosition fromMap(map) => GeolocationPosition(
        latitude: map['lat'] + .0,
        longitude: map['lon'] + .0,
      );
}


lib/src/network.dart
import 'package:dio/dio.dart';

class RequestsHandler {
  late final Dio _dio;

  RequestsHandler(
    String baseUrl, {
    Map<String, dynamic>? headers,
    ResponseType? responseType,
  }) {
    _dio = Dio(BaseOptions(
      baseUrl: baseUrl,
      headers: headers,
      responseType: responseType ?? ResponseType.json,
    ));
  }

  Future<NetworkResponse> getRequest(String path) async {
    final res = await _dio.get(path);
    return NetworkResponse(res.data, res.statusCode);
  }

  Future<NetworkResponse> postRequest(String path, [dynamic body]) async {
    late final Response res;
    if (body != null) {
      res = await _dio.post(path, data: body);
    } else {
      res = await _dio.post(path);
    }
    return NetworkResponse(res.data, res.statusCode);
  }

  Future<NetworkResponse> putRequest(String path, [dynamic body]) async {
    late final Response res;
    if (body != null) {
      res = await _dio.put(path, data: body);
    } else {
      res = await _dio.put(path);
    }
    return NetworkResponse(res.data, res.statusCode);
  }

  Future<NetworkResponse> deleteRequest(String path) async {
    final res = await _dio.delete(path);
    return NetworkResponse(res.data, res.statusCode);
  }
}

class NetworkResponse {
  final int? status;
  final dynamic data;

  NetworkResponse(this.data, [this.status]);
}


example/variscite_dart_example.dart
import 'package:variscite_dart/variscite_dart.dart';
import 'package:geojson_vi/geojson_vi.dart';

void main() async {
  // Create new API instance
  final api = VarisciteApi();

  // Creating new group and user
  final group = await api.createNewGroup(
    GroupCreationBody(name: 'New awesome group', passcode: 'qwerty'),
  );
  final user = await api.loginUsingInviteCode(
    group.inviteCode,
    UserLoginBody(name: 'Admin', passcode: 'qwerty'),
  );
  api.setToken(user.token);
  final groupInfo = await api.getCurrentGroupInfo();

  print('Group "${groupInfo.name}"');

  // Create new structure
  final feature = {
    'type': 'FeatureCollection',
    'features': [
      {
        'type': 'Feature',
        'properties': {},
        'geometry': {
          'type': 'Polygon',
          'coordinates': [
            [
              [34.453125, 54.007768],
              [40.78125, 54.007768],
              [40.78125, 57.326521],
              [34.453125, 57.326521],
              [34.453125, 54.007768],
            ],
          ],
        },
      },
    ],
  };
  final collectionGeoJSON = GeoJSONFeatureCollection.fromMap(feature);
  final newStruct = await api.addNewStructure(collectionGeoJSON);

  print('Structure id: ${newStruct.id}');

  // Delete the group after all
  await api.deleteCurrentGroup();
}
